bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/discovery/TouchWandControllerDiscoveryService.java

```diff
@@ -0,0 +1,146 @@
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.touchwand.internal.TouchWandBindingConstants;
+import org.openhab.binding.touchwand.internal.TouchWandBridgeHandler;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link TouchWandControllerDiscoveryService} Discovery service for Touchwand Controllers.
+ *
+ * @author Roie Geron - Initial contribution
+ */
+@Component(service = DiscoveryService.class, immediate = false, configurationPid = "discovery.touchwand")
+@NonNullByDefault({})
+public class TouchWandControllerDiscoveryService extends AbstractDiscoveryService {
+
+    private static final int SEARCH_TIME_SEC = 2;
+    private static final int TOUCHWAND_BCAST_PORT = 35000;
+    private final Logger logger = LoggerFactory.getLogger(TouchWandControllerDiscoveryService.class);
+
+    private @Nullable Thread socketReceiveThread = null;
+    private @Nullable DatagramSocket listenSocket;
+
+    public TouchWandControllerDiscoveryService() {
+        super(TouchWandBridgeHandler.SUPPORTED_THING_TYPES, SEARCH_TIME_SEC, true);
+        removeOlderResults(getTimestampOfLastScan());
+        try {
+            listenSocket = new DatagramSocket(TOUCHWAND_BCAST_PORT);
+        } catch (SocketException e) {
+            logger.warn("SocketException {}", e.getMessage());
+        }
+    }
+
+    @Override
+    protected void startScan() {
+        runReceiveThread(listenSocket);
+    }
+
+    @Override
+    protected synchronized void stopScan() {
+        super.stopScan();
+    }
+
+    @Override
+    public void deactivate() {
+        if (socketReceiveThread != null) {
+            socketReceiveThread.interrupt();
+            socketReceiveThread = null;
+        }
+        if (listenSocket != null) {
+            listenSocket.close();
+        }
+        super.deactivate();
+    }
+
+    private void addDeviceDiscoveryResult(String label, String ip) {
+        String id = ip.replaceAll("\\.", "");
+        ThingUID thingUID = new ThingUID(THING_TYPE_BRIDGE, id);
+        Map<String, Object> properties = new HashMap<>();
+        properties.put("label", label);
+        properties.put("ipAddress", ip);
+        // @formatter:off
+        logger.debug("Add new Bridge label:{} id {} ",label, id);
+        thingDiscovered(DiscoveryResultBuilder.create(thingUID)
+                .withThingType(THING_TYPE_BRIDGE)
+                .withLabel(label)
+                .withProperties(properties)
+                .withRepresentationProperty("ipAddress")
+                .build()
+        );
+        // @formatter:on
+    }
+
+    protected void runReceiveThread(DatagramSocket socket) {
+        socketReceiveThread = new ReceiverThread(socket);
+        socketReceiveThread.setName(TouchWandBindingConstants.BINDING_ID);
+        socketReceiveThread.start();

```

Comment:
See above. You can store it to a local variable to eliminate the compiler warning.

Reply:
@fwolter 
Done some acrobatics to eliminate all warnings in @NonNullByDefault and all compiler warnings 
in TouchWandControllerDiscoveryService.java
Thread name been changed to match new naming convention 

regarding  unitObj.get("type")  , the "type" is currently always received in the response and  it is very stable . I will look to add some checks , but at this point I prefer to leave it as is if it is ok